### ISSUE TITLE
[FIXED]: ValueError "Input 0 of layer sequential is incompatible with the layer: Expected expected min_ndim=4, found ndim=3." Did Incorrect Loss Function Defined with `tff.learning.build_federated_averaging_process` given model_fn Callable[] tf.TensorShape Create Error for InputTensor Shape Check?

### ISSUE THREAD

**Issue by ferasbg** [tensorflow/federated#1207]
**Describe the bug**
The bug is specifically with the issue with the relation of the defined argument for the input_spec defined for the `model_fn` callable passed in the parameters for `tff.learning.build_federated_averaging_process(model_fn, client_optimizer_fn, server_optimizer_fn)` given the `tensorflow_federated.python.core.templates.iterative_process.IterativeProcess` object. I noticed that when I used the incorrect loss function type, specifically the `tf.keras.losses.CategoricalCrossentropy
` loss function, and the correct `input_shape` for the data passed into the input layer of the plaintext keras model, it was giving me consistent issues of . I wanted to address that the IterativeProcess object was successfully created when I changed it to SparseCategoricalCrossEntropy instead, which was the correct loss function for the plaintext model which was ignored in the plaintext model training, but created serious issues when I was defining the tensor shape given the type specifications required for input data for tensorflow federated. 

**Environment (please complete the following information):**
* OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04 Focal Fossa
* Python package versions (e.g., TensorFlow Federated, TensorFlow): Tensorflow 2.4.1, and Tensorflow Federated 0.18.0. 
* Python version: Python 3.8.5.
* Bazel version (if building from source): None
* CUDA/cuDNN version: None
* What TensorFlow Federated execution stack are you using? I am going to define the args for these functions. I am using `tff.learning.build_federated_averaging_process(model_fn, client_optimizer_fn, server_optimizer_fn)`, as well as 
```
input_spec = collections.OrderedDict(
        x=tf.TensorSpec([None, 32, 32, 3], tf.float32),
        y=tf.TensorSpec([None, 1], tf.int32)
    ), 
```
and `tff.learning.from_keras_model(model, input_spec=input_spec, loss=tf.keras.losses.SparseCategoricalCrossentropy(), metrics=[tf.keras.metrics.SparseCategoricalAccuracy()])`

**Expected behavior**
Given that I am using the CIFAR-100 dataset, the `input_spec` passed in the plaintext to tensorflow federated model converter had incorrectly defined the `tf.TensorSpec` shapes, and I am unsure on how the loss_function is evaluated given just plain `CategoricalCrossEntropy()`, if I had wished to use this loss function instead to evaluate possibly multiple classes. But I do not think the loss function is largely related to the problem, but was unsure on whether there is a native function going on that is error checking for the input_layer given the loss function and how the error was evaluated for the input tensor given incorrect loss function, and whether these are unrelated, and there's another component that I am missing.

**Additional context**
It is important to note that this has already been debugged by changing the loss function and metrics argument passed in the `tff.learning.from_keras_model` function in `model_fn`, but I am unsure on exactly whether the fix in correcting my loss function removed the error as well for the TensorSpec being read. Why was this the case? Perhaps, interpretability errors were sort of apparent since this error was more ambiguous. 

Just note that I am putting this **already resolved issue under bug report since I think it would be optimal to see more interpretability since the solution was sort of counter-intuitive** for me, and noted that it would be useful to hear some feedback as well on this. 

Thank you so much.

**Comment by ZacharyGarrett** [comment]
Thank you for the report!

To confirm: when using the incorrect loss function (in this case `tf.keras.losses.CategoricalCrossentropy` versus `tf.keras.losses.SparseCategoricalCrossnetropy`), the `ValueError` specifying that the tensor dimensions don't work doesn't help the programmer identify and debug the issue?

Any chance we could attach a minimal python snippet that reproduces the error? Including a copy of the error logs and stacktrace would also be extremely helpful.
